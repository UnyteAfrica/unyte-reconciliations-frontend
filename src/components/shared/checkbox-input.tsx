import camelcase from "camelcase";
import { cx } from "class-variance-authority";
import { ChangeEventHandler, MouseEventHandler } from "react";
import { FaCheck } from "react-icons/fa";
import { twMerge } from "tailwind-merge";

type CheckboxInputProps = {
  label: string;
  onClick?: MouseEventHandler<HTMLDivElement>;
  onChange: ChangeEventHandler<HTMLInputElement>;
  name: string;
  checked: boolean;
  className?: string;
};

export const CheckboxInput: React.FC<CheckboxInputProps> = ({
  label,
  checked,
  name,
  onClick,
  onChange,
  className,
}) => {
  return (
    <div className={twMerge("flex items-center mb-4 h-8", className)}>
      <div className="mr-2 flex justify-center items-center">
        <input
          id={camelcase(label)}
          type="checkbox"
          name={name}
          className={cx("w-5 h-5", checked && "hidden")}
          onChange={onChange}
          checked={checked || false}
        />
        <div
          className={twMerge(
            "bg-white border border-mPrimary relative rounded hidden justify-center items-center w-[22px] h-[22px]",
            checked && "flex"
          )}
          onClick={onClick}
        >
          <FaCheck className="absolute -top-1 left-[2px] text-2xl text-mPrimary" />
        </div>
      </div>
      <label htmlFor={camelcase(label)}>{label}</label>
    </div>
  );
};
