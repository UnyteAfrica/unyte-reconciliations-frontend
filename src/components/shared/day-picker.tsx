import moment, { Moment } from "moment";
import { ComponentProps } from "react";
import { ClassNameValue, twMerge } from "tailwind-merge";
import "react-datetime/css/react-datetime.css";
import Datetime from "react-datetime";

type DayPickerProps = {
  className?: ClassNameValue;
  date?: Moment;
} & ComponentProps<typeof Datetime>;

export const DayPicker: React.FC<DayPickerProps> = ({
  className,
  date,
  ...props
}) => {
  return (
    <Datetime
      dateFormat="DD/MM"
      initialViewMode="days"
      value={date}
      className={twMerge(
        "border border-[#ccc] rounded-lg p-2 font-inter font-semibold",
        className
      )}
      renderInput={(props, openCalendar, _) => {
        return (
          <div
            className="w-[150px]"
            onClick={() => {
              openCalendar();
            }}
          >
            <input
              {...props}
              className="w-[150px] text-center cursor-pointer"
              value={date?.format("MMMM DD, YYYY")}
            />
          </div>
        );
      }}
      {...props}
    />
  );
};

const YEAR_1999 = moment("2000");
const endOfCurrentDay = moment().endOf("day");

type RangeDayPickerProps = {
  className?: ClassNameValue;
  startDay?: Moment;
  endDay?: Moment;
  onStartDayChange?: ((value: moment.Moment | string) => void) | undefined;
  onEndDayChange?: ((value: moment.Moment | string) => void) | undefined;
} & ComponentProps<typeof Datetime>;

export const RangeDayPicker: React.FC<RangeDayPickerProps> = ({
  startDay,
  endDay,
  onStartDayChange,
  onEndDayChange,
}) => {
  return (
    <div className="flex flex-col min-[450px]:flex-row">
      <DayPicker
        className="sm:rounded-r-none"
        date={startDay}
        onChange={onStartDayChange}
        isValidDate={(currDate) =>
          currDate.isAfter(YEAR_1999) && currDate.isBefore(endOfCurrentDay)
        }
      />
      <DayPicker
        className="sm:rounded-l-none"
        date={endDay}
        onChange={onEndDayChange}
        isValidDate={(currDate: Moment) =>
          currDate.endOf("day").isAfter(startDay) &&
          currDate.startOf("day").isBefore(endOfCurrentDay)
        }
      />
    </div>
  );
};
