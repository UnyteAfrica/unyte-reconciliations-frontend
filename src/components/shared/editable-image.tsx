import React, { ComponentPropsWithoutRef, useRef, useState } from "react";
import { MdEdit } from "react-icons/md";
import { twMerge } from "tailwind-merge";
import { Loader } from "../loader";

export type EditableProfileImageProps = ComponentPropsWithoutRef<"input"> & {
  extImageUrl?: string;
  error?: string;
  errorClassName?: string;
  initials: string;
  isLoading?: boolean;
  onImageChange: (image: File) => void;
};

export const EditableProfileImage: React.FC<EditableProfileImageProps> = ({
  extImageUrl,
  onImageChange,
  error,
  errorClassName,
  initials,
  isLoading,
}) => {
  const [imageUrl, setImageUrl] = useState(extImageUrl ?? "");
  const imageInputRef = useRef<HTMLInputElement>(null);

  if (isLoading)
    return (
      <div className="w-32 h-32 bg-[#e0e0e0] rounded-full flex justify-center items-center my-4 relative lg:shrink-0 lg:my-0">
        <Loader className="mx-auto w-16 h-16" />
      </div>
    );
  return (
    <>
      <div className="relative inline-block">
        <input
          ref={imageInputRef}
          className="hidden"
          type="file"
          onChange={(e) => {
            if (e.target.files) {
              const file = e.target.files[0];
              setImageUrl(URL.createObjectURL(file));
              onImageChange(file);
            }
          }}
          accept="image/png, image/jpeg, image/webp, image/jpg"
        />
        <div className="relative inline-block">
          {!error && imageUrl ? (
            <img
              className="w-32 h-32 rounded-full flex my-4 object-cover lg:shrink-0 lg:my-0"
              src={imageUrl}
              alt=""
            />
          ) : (
            <div className="w-32 h-32 bg-[#e0e0e0] rounded-full flex justify-center items-center my-4 relative lg:shrink-0 lg:my-0">
              <em className="not-italic font-semibold text-3xl">{initials}</em>
            </div>
          )}
          <div
            className="w-8 h-8 rounded-full bg-black text-white absolute bottom-2 right-2 flex justify-center items-center cursor-pointer"
            onClick={() => {
              imageInputRef?.current?.click();
            }}
          >
            <MdEdit className="text-xl" />
          </div>
        </div>

        <p
          className={twMerge(
            "font-sora text-lg block font-bold  mt-2 text-red-600",
            errorClassName
          )}
        >
          {error && error}
        </p>
      </div>
    </>
  );
};
