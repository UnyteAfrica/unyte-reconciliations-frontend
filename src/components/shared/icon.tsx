import { ComponentPropsWithoutRef, FC } from "react";
import { clsx } from "clsx";

import logo from "@/assets/Icons/UnyteLogo.svg";

const iconTypes = {
  logo,
};

export type IconType = keyof typeof iconTypes;

type IconProps = Omit<ComponentPropsWithoutRef<"img">, "src" | "alt"> & {
  alt?: string;
  type: IconType;
  className?: string;
};

export const Icon: FC<IconProps> = ({ type, alt, className, ...props }) => {
  return (
    <img
      {...props}
      src={iconTypes[type]}
      alt={alt ? alt : ""}
      className={clsx("", className ? className : "")}
    />
  );
};
