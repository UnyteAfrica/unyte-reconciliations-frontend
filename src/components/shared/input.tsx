"use client";

import { cx } from "class-variance-authority";
import {
  ComponentPropsWithoutRef,
  MouseEventHandler,
  ReactNode,
  forwardRef,
  useImperativeHandle,
  useRef,
  useState,
} from "react";
import { twMerge } from "tailwind-merge";
import camelcase from "camelcase";
import { FaRegEye } from "react-icons/fa";
import { Image } from "./image";

export type CustomInputProps = ComponentPropsWithoutRef<"input"> & {
  label?: string;
  labelClassName?: string;
  containerClassName?: string;
  error?: string;
  errorClassName?: string;
  after?: ReactNode;
  optional?: boolean;
};
export type ImageInputProps = ComponentPropsWithoutRef<"input"> & {
  label?: string;
  labelClassName?: string;
  extImageUrl?: string;
  containerClassName?: string;
  error?: string;
  errorClassName?: string;
  onImageChange: (image: File) => void;
};

export type CustomSelectProps = ComponentPropsWithoutRef<"select"> & {
  label?: string;
  labelClassName?: string;
  containerClassName?: string;
  selectWrapperClassname?: string;
  error?: string;
  errorClassName?: string;
  placeholder?: string;
  after?: ReactNode;
  options: { option: string; value: string }[] | string[];
};

type CustomTextAreaProps = ComponentPropsWithoutRef<"textarea"> & {
  label?: string;
  labelClassName?: string;
  error?: string;
  errorClassName?: string;
};

type PasswordInputProps = Omit<CustomInputProps, "label"> & {
  inputClassname?: string;
  error?: string;
  label?: string;
};

const isStringArray = (
  obj: { option: string; value: string }[] | string[]
): obj is string[] => {
  return typeof obj[0] == "string";
};

export const CustomInput = forwardRef<HTMLInputElement, CustomInputProps>(
  function CustomInput(props, ref) {
    const {
      placeholder,
      type,
      className,
      label = "",
      labelClassName = "",
      containerClassName = "",
      error,
      errorClassName,
      after,
      optional,
      ...rest
    } = props;
    const inputId = camelcase(label);

    return (
      <div className={twMerge("", containerClassName)}>
        {label ? (
          <>
            <label
              className={twMerge(
                "text-lg text-[#333] leading-[24px] mb-2 block",
                labelClassName
              )}
              htmlFor={inputId}
            >
              {label}{" "}
              {optional && (
                <span className="not-italic text-sm text-[#333]">
                  (optional)
                </span>
              )}
            </label>
          </>
        ) : null}

        <input
          ref={ref}
          className={twMerge(
            "block text-lg font-sora w-full px-3 py-[16px] h-[72px] border border-[#424242] rounded-lg outline-none focus:border-black",
            className
          )}
          id={inputId}
          type={type}
          placeholder={placeholder}
          {...rest}
        />
        <p className={twMerge("text-sm text-[#333]", errorClassName)}>
          {error && error}
        </p>
        {/* {after && after} */}
      </div>
    );
  }
);

export const CustomSelect = forwardRef<HTMLSelectElement, CustomSelectProps>(
  function CustomSelect(props, ref) {
    const {
      className,
      label = "",
      labelClassName = "",
      containerClassName = "",
      selectWrapperClassname,
      error,
      errorClassName,
      placeholder,
      after,
      options,
      ...rest
    } = props;
    const inputId = camelcase(label);
    return (
      <div className={twMerge("", containerClassName)}>
        {label ? (
          <label
            className={twMerge(
              "font-sora font-semibold text-[16px] leading-[20px] mb-2 block",
              labelClassName
            )}
            htmlFor={inputId}
          >
            {label}
          </label>
        ) : null}

        <div
          className={twMerge(
            "border-[#B5B3B3] border inline-block rounded-lg overflow-hidden relative w-[156px] h-[54px]",
            selectWrapperClassname
          )}
        >
          <select
            ref={ref}
            className={twMerge(
              "block text-[14px] font-sora w-full px-3 py-[16px] border border-[#424242] rounded-lg outline-none",
              className
            )}
            id={inputId}
            {...rest}
          >
            <option disabled value="">
              {placeholder
                ? placeholder
                : ` Select ${label ? `a ${label.toLowerCase()}` : "an option"}`}
            </option>
            {isStringArray(options)
              ? options.map((option, idx) => (
                  <option value={option} key={idx}>
                    {option}
                  </option>
                ))
              : options.map((option, idx) => (
                  <option value={option.value} key={idx}>
                    {option.option}
                  </option>
                ))}
          </select>
        </div>
        <p className={twMerge("font-sora text-sm text-[#333]", errorClassName)}>
          {error && error}
        </p>
      </div>
    );
  }
);

export const CustomTextArea = forwardRef<
  HTMLTextAreaElement,
  CustomTextAreaProps
>(function CustomTextArea(props, ref) {
  const {
    placeholder,
    className,
    label = "",
    labelClassName = "",
    error = "",
    errorClassName,
    ...rest
  } = props;
  const inputId = camelcase(label);
  return (
    <div className="mb-8">
      {label ? (
        <label
          className={twMerge(
            "font-sora font-semibold text-[16px] leading-[20px] mb-2 block",
            labelClassName
          )}
          htmlFor={inputId}
        >
          {label}
        </label>
      ) : null}

      <textarea
        className={twMerge(
          "block text-[14px] w-full px-[3px] transition rounded-lg duration-300 focus:border-black",
          className
        )}
        ref={ref}
        placeholder={placeholder}
        {...rest}
      />
      <p className={twMerge("font-sora text-sm text-[#333]", errorClassName)}>
        {error && error}
      </p>
    </div>
  );
});

export const PasswordInput = forwardRef<HTMLInputElement, PasswordInputProps>(
  function PasswordInput(props, ref) {
    const {
      className,
      inputClassname = "",
      labelClassName = "",
      label = "Password",
      onChange,
      value,
      error,
      ...rest
    } = props;
    const [isPasswordShown, setIsPasswordShown] = useState(false);

    const togglePasswordShown: MouseEventHandler<HTMLButtonElement> = (e) => {
      e.preventDefault();
      setIsPasswordShown((oldIsPasswordShown) => !oldIsPasswordShown);
    };

    const inputId = camelcase(label);

    return (
      <div>
        <label
          className={twMerge(
            "text-lg text-[#333] leading-[24px] mb-2 block",
            labelClassName
          )}
          htmlFor={inputId}
        >
          {label}
        </label>

        <div
          className={twMerge(
            "flex relative items-center w-[100%] border border-[#424242] rounded-lg  transition duration-300 focus-within:border-black",
            className
          )}
        >
          <CustomInput
            type={isPasswordShown ? "text" : "password"}
            placeholder="********"
            className={twMerge(
              "mb-0 border-0 w-[90%] focus:border-transparent focus:border-b-0",
              inputClassname
            )}
            containerClassName="w-full"
            onChange={onChange}
            value={value}
            id={inputId}
            ref={ref}
            {...rest}
          />

          <button
            onClick={togglePasswordShown}
            className={cx(
              "absolute top-[50%] -translate-y-[50%] right-3",
              isPasswordShown &&
                "after:content-[''] after:absolute after:left-2 after:top-0 after:h-full after:w-[2px] after:bg-[#595959] after:rotate-45"
            )}
          >
            <FaRegEye width={20} height={20} />
          </button>
        </div>
        <p className="font-sora text-sm text-[#333]">{error && error}</p>
      </div>
    );
  }
);

export const ImageInput = forwardRef<HTMLInputElement, ImageInputProps>(
  function ImageInput(props, ref) {
    const {
      placeholder,
      type,
      className,
      label = "",
      labelClassName = "",
      containerClassName = "",
      error,
      errorClassName,
      extImageUrl,
      onImageChange,
      ...rest
    } = props;
    const inputId = camelcase(label);
    const [imageUrl, setImageUrl] = useState(extImageUrl ?? "");
    const imageInputRef = useRef<HTMLInputElement>(null);
    useImperativeHandle(ref, () => imageInputRef.current!, []);
    return (
      <div className={twMerge("", containerClassName)}>
        {label ? (
          <>
            <label
              className={twMerge(
                "font-sora font-semibold text-[16px] leading-[20px] mb-2 block",
                labelClassName
              )}
              htmlFor={inputId}
            >
              {label}
            </label>
            <p className="my-0 mb-2 text-sm">
              Tap the box to select the image*
            </p>
          </>
        ) : null}

        <input
          ref={imageInputRef}
          className={twMerge(
            "text-[14px] font-sora w-full px-3 py-[16px] border border-[#424242] rounded-lg outline-none focus:border-black hidden",
            className
          )}
          id={inputId}
          type="file"
          placeholder={placeholder}
          onChange={(e) => {
            if (e.target.files) {
              const file = e.target.files[0];
              setImageUrl(URL.createObjectURL(file));
              onImageChange(file);
            }
          }}
          accept="image/png, image/jpeg, image/webp, image/jpg"
          {...rest}
        />
        <div
          className="w-full aspect-[282/131] border border-[#838080] flex justify-center items-center rounded-lg"
          onClick={() => {
            imageInputRef?.current?.click();
          }}
        >
          {imageUrl ? (
            <Image
              src={imageUrl}
              className="w-full h-full object-cover rounded-lg"
              alt="Image preview"
              height={160}
              width={328}
            />
          ) : (
            <p>Select Image File</p>
          )}
        </div>
        <p className={twMerge("font-sora text-sm text-[#333]", errorClassName)}>
          {error && error}
        </p>
      </div>
    );
  }
);

type DateInputProps = {
  containerClassName?: string;
  date: Date;
  onDateChange: (date: Date) => void;
};

export const DateInput = forwardRef<HTMLInputElement, DateInputProps>(
  function DateInput(props, _) {
    const { containerClassName, date, onDateChange } = props;
    const inputRef = useRef<HTMLInputElement>(null);

    const formattedDate = date.toLocaleDateString("en-US", {
      year: "numeric",
      month: "long",
      day: "numeric",
      // weekday: "short", // Optional, to include weekday (e.g., 'Sun')
    });

    const inputValue = date.toISOString().split("T")[0];

    return (
      <button
        className={twMerge(
          "px-4 py-2 border rounded-md relative",
          containerClassName ?? ""
        )}
        onClick={() => inputRef.current?.showPicker()}
      >
        {formattedDate}
        <input
          type="date"
          placeholder="date"
          className="absolute top-[45px] left-0 invisible opacity-0 h-0 w-0"
          value={inputValue}
          ref={inputRef}
          onChange={(e) => {
            const dateVal = e.target.valueAsDate;
            if (dateVal) {
              onDateChange(dateVal);
            }
          }}
        />
      </button>
    );
  }
);
