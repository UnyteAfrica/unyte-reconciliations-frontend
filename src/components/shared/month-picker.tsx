import moment, { Moment } from "moment";
import React, { ComponentProps } from "react";
import { ClassNameValue, twMerge } from "tailwind-merge";
import "react-datetime/css/react-datetime.css";
import Datetime from "react-datetime";

type MonthPickerProps = {
  className?: ClassNameValue;
  date?: Moment;
} & ComponentProps<typeof Datetime>;

export const MonthPicker: React.FC<MonthPickerProps> = ({
  className,
  date,
  ...props
}) => {
  return (
    <Datetime
      dateFormat="MM"
      initialViewMode="months"
      value={date}
      className={twMerge(
        "border border-[#ccc] rounded-lg p-2 font-inter font-semibold",
        className
      )}
      renderInput={(props, openCalendar, _) => {
        return (
          <div
            className="w-[100px]"
            onClick={() => {
              openCalendar();
            }}
          >
            <input
              {...props}
              className="w-[100px] text-center cursor-pointer"
              value={date?.format("MMMM")}
            />
          </div>
        );
      }}
      {...props}
    />
  );
};

const YEAR_1999 = moment("2000");
const endOfCurrentMonth = moment().endOf("month");

type RangeMonthPickerProps = {
  className?: ClassNameValue;
  startMonth?: Moment;
  endMonth?: Moment;
  onStartMonthChange?: ((value: moment.Moment | string) => void) | undefined;
  onEndMonthChange?: ((value: moment.Moment | string) => void) | undefined;
} & ComponentProps<typeof Datetime>;

export const RangeMonthPicker: React.FC<RangeMonthPickerProps> = ({
  startMonth,
  endMonth,
  onStartMonthChange,
  onEndMonthChange,
}) => {
  return (
    <div className="flex flex-col min-[450px]:flex-row">
      <MonthPicker
        className="sm:rounded-r-none"
        date={startMonth}
        onChange={onStartMonthChange}
        isValidDate={(currDate) =>
          currDate.isAfter(YEAR_1999) && currDate.isBefore(endOfCurrentMonth)
        }
      />
      <MonthPicker
        className="sm:rounded-l-none"
        date={endMonth}
        onChange={onEndMonthChange}
        isValidDate={(currDate: Moment) =>
          currDate.isAfter(startMonth) && currDate.isBefore(endOfCurrentMonth)
        }
      />
    </div>
  );
};
