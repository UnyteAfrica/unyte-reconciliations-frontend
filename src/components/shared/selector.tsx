import { cx } from "class-variance-authority";
import { Fragment, useEffect, useRef, useState } from "react";
import { BiChevronDown, BiChevronUp } from "react-icons/bi";

type SelectorProps = {
  containerClassName?: string;
  selectorClassName?: string;
  options: readonly string[];
  value: string;
  onChange: (val: string) => void;
};

export const Selector: React.FC<SelectorProps> = ({
  containerClassName,
  selectorClassName,
  options,
  value,
  onChange,
}) => {
  const [isSelectorMenuOpen, setIsSelectorMenuOpen] = useState(false);
  const containerRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    document.querySelector("body")?.addEventListener("click", (e) => {
      if (!containerRef.current?.contains(e.target as Node)) {
        setIsSelectorMenuOpen(false);
      }
    });
  }, []);

  return (
    <div className={cx("relative", containerClassName)} ref={containerRef}>
      <button
        className="px-4 py-2 text-[#333333] rounded-md border font-semibold w-full"
        onClick={() =>
          setIsSelectorMenuOpen((isSelectorMenuOpen) => !isSelectorMenuOpen)
        }
      >
        <div className="space-x-2 flex flex-row items-center justify-between">
          <span
            className={cx("text-base", selectorClassName)}
            title="selectBtn"
          >
            {value}
          </span>
          {isSelectorMenuOpen ? <BiChevronUp /> : <BiChevronDown />}
        </div>
      </button>
      <div
        className={cx(
          "absolute z-20 w-full top-[45px] left-0 bg-white border rounded-lg border-[#e5e5e5] max-h-0 transition-all duration-300 overflow-hidden opacity-0",
          isSelectorMenuOpen && "max-h-[250px] opacity-100 overflow-y-auto"
        )}
      >
        {options.map((option, idx) => (
          <Fragment key={idx}>
            <button
              className="w-full font-inter py-4 font-semibold"
              onClick={() => {
                setIsSelectorMenuOpen(false);
                onChange(option);
              }}
            >
              {option}
            </button>
            <hr />
          </Fragment>
        ))}
      </div>
    </div>
  );
};
