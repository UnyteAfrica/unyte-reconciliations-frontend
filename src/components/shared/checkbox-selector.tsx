import { cx } from "class-variance-authority";
import { Fragment, useEffect, useRef } from "react";
import { BiChevronDown, BiChevronUp } from "react-icons/bi";
import { CheckboxInput } from "./checkbox-input";

type CheckboxSelectorProps = {
  containerClassName?: string;
  selectorClassName?: string;
  selectionMap: Record<string, boolean>;
  options: readonly string[];
  value: string;
  isCheckboxSelectorMenuOpen: boolean;
  setIsCheckboxSelectorMenuOpen: (
    val: boolean | ((val: boolean) => boolean)
  ) => void;
  onChange: (val: string, isSelected: boolean) => void;
  resetSelectedMap: () => void;
};

export const CheckboxSelector: React.FC<CheckboxSelectorProps> = ({
  containerClassName,
  selectorClassName,
  options,
  value,
  onChange,
  selectionMap,
  isCheckboxSelectorMenuOpen,
  setIsCheckboxSelectorMenuOpen,
  resetSelectedMap,
}) => {
  const containerRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    document.querySelector("body")?.addEventListener("click", (e) => {
      if (!containerRef.current?.contains(e.target as Node)) {
        setIsCheckboxSelectorMenuOpen(false);
      }
    });
  }, []);

  return (
    <div
      className={cx("relative", containerClassName)}
      ref={containerRef}
      onClick={(e) => e.stopPropagation()}
    >
      <button
        className="px-4 py-2 text-[#333333] rounded-md border font-semibold w-full"
        onClick={() => {
          setIsCheckboxSelectorMenuOpen(
            (isCheckboxSelectorMenuOpen) => !isCheckboxSelectorMenuOpen
          );
        }}
      >
        <div className="space-x-2 flex flex-row items-center justify-between">
          <span
            className={cx("text-base", selectorClassName)}
            title="selectBtn"
          >
            {value || "None"}
          </span>
          {isCheckboxSelectorMenuOpen ? (
            <BiChevronUp />
          ) : (
            <BiChevronDown
              onClick={(e) => {
                e.stopPropagation();
                setIsCheckboxSelectorMenuOpen(true);
              }}
            />
          )}
        </div>
      </button>
      <div
        className={cx(
          "absolute z-20 w-full top-[45px] left-0 bg-white border rounded-lg border-[#e5e5e5] max-h-0 transition-all duration-300 overflow-hidden opacity-0",
          isCheckboxSelectorMenuOpen &&
            "max-h-[200px] opacity-100 overflow-y-auto"
        )}
      >
        <Fragment>
          <div className="w-full font-inter p-4 font-semibold">
            <button onClick={resetSelectedMap}>Reset</button>
          </div>
          <hr />
        </Fragment>
        {options.map(
          (option, idx) =>
            option && (
              <Fragment key={idx}>
                <div className="w-full font-inter p-4 font-semibold">
                  <CheckboxInput
                    label={option || "None"}
                    className="mb-0"
                    onClick={() => onChange(option, !selectionMap[option])}
                    onChange={(e) => onChange(option, e.target.checked)}
                    checked={selectionMap[option]}
                    name="policy-type"
                  />
                </div>
                <hr />
              </Fragment>
            )
        )}
      </div>
    </div>
  );
};
