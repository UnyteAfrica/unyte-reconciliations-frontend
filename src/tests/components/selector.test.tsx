import { render, screen } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import React from "react";
import { Selector } from "@/components/shared/selector";

function setup(reactNode: React.ReactNode) {
  return {
    user: userEvent.setup(),
    ...render(reactNode),
  };
}

describe("Selector", () => {
  it("renders correctly", () => {
    const tree = render(
      <BrowserRouter>
        <Selector
          options={["Hello", "World", "Foo"]}
          containerClassName="w-full"
          value={"Hello"}
          onChange={(_) => {}}
        />
      </BrowserRouter>
    );
    expect(tree).toMatchSnapshot();
  });

  it("should open dropdown", async () => {
    const { user } = setup(
      <BrowserRouter>
        <Selector
          options={["Hello", "World", "Foo"]}
          containerClassName="w-full"
          value={"Hello"}
          onChange={(_) => {}}
        />
      </BrowserRouter>
    );

    const valueSelector = screen.getByTitle("selectBtn");
    await user.click(valueSelector);
    const dailyElems = await screen.findAllByText("Hello");
    const helloBtn = dailyElems[1];
    const worldBtn = screen.getByText("World");
    const fooBtn = screen.getByText("Foo");

    expect(helloBtn).toBeVisible();
    expect(worldBtn).toBeVisible();
    expect(fooBtn).toBeVisible();
  });

  it("selecting a new period should change the period", async () => {
    const onValueChange = vi.fn();
    const { user } = setup(
      <BrowserRouter>
        <Selector
          options={["Hello", "World", "Foo"]}
          containerClassName="w-full"
          value={"Hello"}
          onChange={(val) => onValueChange(val)}
        />
      </BrowserRouter>
    );

    const valueSelector = screen.getByTitle("selectBtn");
    await user.click(valueSelector);
    const worldBtn = screen.getByText("World");
    await user.click(worldBtn);
    expect(onValueChange).toBeCalledWith("World");
  });
});
