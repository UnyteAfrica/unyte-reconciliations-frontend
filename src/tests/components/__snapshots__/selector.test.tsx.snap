// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Selector > renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="relative w-full"
      >
        <button
          class="px-4 py-2 text-[#333333] rounded-md border font-semibold w-full"
        >
          <div
            class="space-x-2 flex flex-row items-center justify-between"
          >
            <span
              class="text-base"
              title="selectBtn"
            >
              Hello
            </span>
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M16.293 9.293 12 13.586 7.707 9.293l-1.414 1.414L12 16.414l5.707-5.707z"
              />
            </svg>
          </div>
        </button>
        <div
          class="absolute z-20 w-full top-[45px] left-0 bg-white border rounded-lg border-[#e5e5e5] max-h-0 transition-all duration-300 overflow-hidden opacity-0"
        >
          <button
            class="w-full font-inter py-4 font-semibold"
          >
            Hello
          </button>
          <hr />
          <button
            class="w-full font-inter py-4 font-semibold"
          >
            World
          </button>
          <hr />
          <button
            class="w-full font-inter py-4 font-semibold"
          >
            Foo
          </button>
          <hr />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="relative w-full"
    >
      <button
        class="px-4 py-2 text-[#333333] rounded-md border font-semibold w-full"
      >
        <div
          class="space-x-2 flex flex-row items-center justify-between"
        >
          <span
            class="text-base"
            title="selectBtn"
          >
            Hello
          </span>
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M16.293 9.293 12 13.586 7.707 9.293l-1.414 1.414L12 16.414l5.707-5.707z"
            />
          </svg>
        </div>
      </button>
      <div
        class="absolute z-20 w-full top-[45px] left-0 bg-white border rounded-lg border-[#e5e5e5] max-h-0 transition-all duration-300 overflow-hidden opacity-0"
      >
        <button
          class="w-full font-inter py-4 font-semibold"
        >
          Hello
        </button>
        <hr />
        <button
          class="w-full font-inter py-4 font-semibold"
        >
          World
        </button>
        <hr />
        <button
          class="w-full font-inter py-4 font-semibold"
        >
          Foo
        </button>
        <hr />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
